### DISCUSSION TITLE
Uncaught Error: Dynamic require of "fs/promises" is not supported- calling fetch in the client

### DISCUSSION THREAD

**Discussion by juniorforlife** [remix-run/remix#4976] [Q&A]
I'm facing an issue where my Remix app is really slow between page transition. I'm using RESTful api from a separate backend. I'm not sure if it's the api that makes it slow or Remix so I'm experimenting by calling the api in the client but I got this error
`Uncaught Error: Dynamic require of "fs/promises" is not supported`

Does Remix forbid the client to make http request?

Here's my code
```
export async function loader() {
  return null;
}

export default function CoursesPage() {
  const [data, setData] = useState({ courses: [] });

  useEffect(() => {
    fetch("https://mydomain/api/courses")
      .then((response) => response.json())
      .then((data) => setData({ courses: data.data }))
      .catch((error) => {
        throw json(error);
      });
  }, []);
}
```

**Comment by kiliman** [comment]
No, you can use `fetch` or any other library from your client. 

Is the `/api/courses` endpoint a Remix route? Are you importing `fs/promises`? And if so, how? For node packages, it's recommended you import those libs from _*.server_ file and re-export. This way, Remix will make sure not to include any of those in the client bundle.

Anyway, more code would be helpful if you're still stuck.

**Reply by juniorforlife** [reply]
That's all the code I had. `/api/courses` is a RESTful api endpoint written in Laravel (from a different server, not Remix).  My current project uses Remix only for the client. I only use fetch and no `fs/promises`

**Comment by aniravi24** [comment]
Whenever I see this error occur, it's caused by server-side code that's executing on the client. To expand on what @kiliman suggested above, I would do the following:
1. Determine what route is showing the error (if it's in your root one you'll likely see the error on every page).
2. Look through your imports and see if anything is only meant to be used on the server, the brute force way to do this is just start commenting stuff out until the error goes away
3. Create a file that ends with `.server.ts` (or `.server.js` if you're using JS), for example, `zod.server.ts` and export the particular imports you want to use once you've determined what import/package is causing it.
4. Change your imports to import from this `.server.ts` file instead, and then the error should go away and the app should work normally again!
